source/op/pt/comm.cc

```diff
@@ -0,0 +1,219 @@
+#endif
+    using FPTYPE = double;
+    torch::autograd::variable_list saved_variables = ctx->get_saved_variables();
+    torch::Tensor sendlist_tensor = saved_variables[0];
+    torch::Tensor sendproc_tensor = saved_variables[1];
+    torch::Tensor recvproc_tensor = saved_variables[2];
+    torch::Tensor sendnum_tensor = saved_variables[3];
+    torch::Tensor recvnum_tensor = saved_variables[4];
+    torch::Tensor communicator_tensor = saved_variables[5];
+    torch::Tensor nlocal_tensor = saved_variables[6];
+    torch::Tensor nghost_tensor = saved_variables[7];
+
+    torch::Tensor d_local_g1_tensor = grad_output[0];
+
+    int** recvlist = reinterpret_cast<int**>(sendlist_tensor.data_ptr());
+    // swap send and recv here
+    int* recvproc = sendproc_tensor.data_ptr<int>();
+    int* sendproc = recvproc_tensor.data_ptr<int>();
+    int* recvnum = sendnum_tensor.data_ptr<int>();
+    int* sendnum = recvnum_tensor.data_ptr<int>();
+
+    FPTYPE* local_g1 = d_local_g1_tensor.data_ptr<FPTYPE>();
+    int tensor_size = d_local_g1_tensor.size(1);
+    int nswap = sendproc_tensor.size(0);
+
+    int nlocal = nlocal_tensor.item<int>();
+    int nghost = nghost_tensor.item<int>();
+    int ntotal = nlocal + nghost;
+
+    torch::Tensor send_g1_tensor = d_local_g1_tensor;
+
+    int max_recvnum = sendnum_tensor.max().item<int>();
+    auto options = torch::TensorOptions()
+                       .dtype(torch::kFloat64)
+                       .device(d_local_g1_tensor.device());
+    torch::Tensor recv_g1_tensor =
+        torch::empty({max_recvnum, tensor_size}, options);
+    FPTYPE* recv_g1 = recv_g1_tensor.data_ptr<FPTYPE>();
+    FPTYPE* send_g1 = send_g1_tensor.data_ptr<FPTYPE>() + ntotal * tensor_size;
+
+    MPI_Comm world;
+    unpack_communicator(communicator_tensor, world);
+    int me;
+    MPI_Comm_rank(world, &me);
+    MPI_Datatype mpi_type = get_mpi_type<FPTYPE>();
+    MPI_Request request;
+
+    std::string msg;
+
+    int end = ntotal;
+    auto int32_options = torch::TensorOptions().dtype(torch::kInt32);
+    for (int iswap = nswap - 1; iswap >= 0; --iswap) {
+      int nrecv = recvnum[iswap];
+      int nsend = sendnum[iswap];
+
+      torch::Tensor irecvlist;
+      if (nrecv) {
+        irecvlist = torch::from_blob(recvlist[iswap], {nrecv}, int32_options)
+                        .to(d_local_g1_tensor.device());
+      }
+      if (nsend) {
+        send_g1 -= nsend * tensor_size;
+      }
+      if (sendproc[iswap] != me) {
+        if (nrecv) {
+          MPI_Irecv(recv_g1, nrecv * tensor_size, mpi_type, recvproc[iswap], 0,
+                    world, &request);
+        }
+        if (nsend) {
+          MPI_Send(send_g1, nsend * tensor_size, mpi_type, sendproc[iswap], 0,
+                   world);
+        }
+        if (nrecv) {
+          MPI_Wait(&request, MPI_STATUS_IGNORE);
+        }
+      } else {
+        if (nrecv) {
+#ifdef USE_CUDA
+          cudaMemcpy(recv_g1, send_g1, nrecv * tensor_size * sizeof(FPTYPE),
```

Comment:
We may use [`gpuMemcpy`](https://docs.deepmodeling.com/projects/deepmd/en/latest/api_core/define_gpu__cuda_8h_1a4d7a4ae38c7a1fea58f92c95120471f2.html#define-gpumemcpy) to support both CUDA and ROCm.